json4s/src/main/scala/org/http4s/json4s/Json4sInstances.scala

```diff
@@ -17,17 +17,19 @@ import org.typelevel.jawn.support.json4s.Parser
 object CustomParser extends Parser(useBigDecimalForDouble = true, useBigIntForLong = true)
 
 trait Json4sInstances[J] {
-  implicit def jsonDecoder[F[_]](implicit F: Sync[F]): EntityDecoder[F, JValue] =
+  implicit def jsonDecoder[F[_]](implicit F: Concurrent[F]): EntityDecoder[F, JValue] =
     jawn.jawnDecoder(F, CustomParser.facade)
 
-  def jsonOf[F[_], A](implicit reader: Reader[A], F: Sync[F]): EntityDecoder[F, A] =
+  def jsonOf[F[_], A](implicit reader: Reader[A], F: Concurrent[F]): EntityDecoder[F, A] =
     jsonDecoder.flatMapR { json =>
       DecodeResult(
-        F.delay(reader.read(json))
-          .map[Either[DecodeFailure, A]](Right(_))
-          .recover { case e: MappingException =>
-            Left(InvalidMessageBodyFailure("Could not map JSON", Some(e)))
-          })
+        F.pure {
+          try Right(reader.read(json))
+          catch {
+            case e: Exception => Left(InvalidMessageBodyFailure("Could not map JSON", Some(e)))
```

Comment:
I'd go with `NonFatal(e)` instead of `e: Exception` here.

Reply:
https://github.com/http4s/http4s/pull/3885/commits/44c0b8f9fcf4f161f584356a7b21a3df4c0a8590